example/ipsec_offload/odp_ipsec_offload.c

```diff
@@ -0,0 +1,896 @@
+			if (ipsec_entry)
+				flow->out_port.sa = ipsec_entry->sa;
+			else
+				flow->out_port.sa = ODP_IPSEC_SA_INVALID;
+			flow->next = NULL;
+			/*Insert new flow into flow cache table*/
+			odp_route_flow_insert_in_bucket(flow,
+							&flow_table[hash &
+							(bucket_count - 1)]);
+			goto do_opt;
+		} else {
+			EXAMPLE_DBG("No flow match found. Packet is dropped.\n");
+			odp_packet_free(pkt);
+			return PKT_DROP;
+		}
+	}
+	return ret;
+}
+
+/**
+ * Packet Processing - Input IPsec packet classification
+ *
+ * Verify the received packet has IPsec headers,
+ * if so issue ipsec request else skip.
+ *
+ * @param pkt   Packet to classify
+ *
+ * @return PKT_CONTINUE if done else PKT_POSTED
+ */
+static
+pkt_disposition_e do_ipsec_in_classify(odp_packet_t pkt)
+{
+	odp_ipsec_op_param_t params;
+
+	if (!odp_packet_has_ipsec(pkt))
+		return PKT_CONTINUE;
+
+	/* Initialize parameters block */
+	params.pkt = &pkt;
+	params.num_pkt = 1;
+	params.num_sa = 0;
+	params.num_opt = 0;
+	params.opt = NULL;
+
+	/* Issue ipsec request */
+	if (odp_unlikely(odp_ipsec_in_enq(&params) < 0)) {
+		EXAMPLE_DBG("Unable to in enqueue\n");
+		odp_packet_free(pkt);
+		return PKT_DROP;
+	}
+	return PKT_POSTED;
+}
+
+/**
+ * Packet IO worker thread
+ *
+ * Loop calling odp_schedule to obtain packet from the two sources,
+ * and continue processing the packet.
+ *
+ *  - Input interfaces (i.e. new work)
+ *  - Per packet ipsec API completion queue
+ *
+ * @param arg  Required by "odph_linux_pthread_create", unused
+ *
+ * @return NULL (should never return)
+ */
+static
+int pktio_thread(void *arg EXAMPLE_UNUSED)
+{
+	int thr;
+	odp_packet_t pkt;
+	odp_pktout_queue_t out_queue;
+	odp_out_entry_t	*out_port;
+	odp_event_t ev = ODP_EVENT_INVALID;
+
+	thr = odp_thread_id();
+
+	printf("Pktio thread [%02i] starts\n", thr);
+	odp_barrier_wait(&sync_barrier);
+
+	/* Loop packets */
+	for (;;) {
+		pkt_disposition_e rc;
+
+		ev = odp_schedule(NULL, ODP_SCHED_WAIT);
+		/* Use schedule to get event from any input queue */
+		/* Determine new work versus completion */
+		if (ODP_EVENT_PACKET == odp_event_type(ev)) {
+			pkt = odp_packet_from_event(ev);
+
+			rc = do_input_verify(pkt);
+			if (odp_unlikely(rc))
+				continue;
+
+			rc = do_ipsec_in_classify(pkt);
+			if (rc)
+				continue;
+
+			rc = do_route_fwd_db(pkt);
+			if (rc)
+				continue;
+
+			out_port = (odp_out_entry_t *)odp_packet_user_ptr(pkt);
+			out_queue = (odp_pktout_queue_t)out_port->pktout;
+
+			if (odp_unlikely(odp_pktout_send(out_queue, &pkt, 1)
+							 < 0))
+				odp_packet_free(pkt);
+
+		} else if (ODP_EVENT_IPSEC_RESULT == odp_event_type(ev)) {
+			odp_ipsec_op_result_t result;
+			odp_ipsec_packet_result_t res;
+			odph_ethhdr_t	*eth;
+			odp_packet_t out_pkt;
+
+			result.pkt = &out_pkt;
+			result.res = &res;
+
+			if (odp_unlikely(odp_ipsec_result(&result, ev) < 0)) {
+				EXAMPLE_DBG("Error Event\n");
+				odp_event_free((odp_event_t)ev);
+				continue;
+			}
```

Comment:
According to current API, there should be a call to odp_event_free() here.

My proposal removes need for it if odp_ipsec_result returns 0.